### ISSUE TITLE
Naming the third nullness annotation (the one that matches the behavior of unannotated code)

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#32]
We have seemed fairly comfortable with `@Nullable` and `@NotNull` (the latter being a shorthand for `@NotNullable`), but oy, the third one.

The word "legacy" should be involved; we have never come up with something more appropriate. It is the nullness that all Java types have by default *before* someone goes through and annotates. Admittedly, people will probably use it (rarely) for special cases with non-legacy code, but we think we want to at least discourage that.

I think we also don't want to talk about it as being more closely related to `@Nullable` than to `@NotNull`; its relationship to these two should be symmetric. It represents indecision between the two. Conservative analysis treats it more like `@Nullable` when assigning from, but more like `@NotNull` when assigning to; lenient analysis would have it the other way around.

Names

`@LegacyNull` / `@DefaultLegacyNull`
`@LegacyNullness` / `@DefaultLegacyNullness`
`@NullableLegacy` / `@DefaultNullableLegacy` (sorts/groups better)
`@NullUnchecked` / `@DefauiltNullUnchecked` (same)
`@SchroedingersNull`

**Comment by mernst** [comment]
Thanks for this background and rationale.  That is helpful.

I think we should use different names for the different cases, rather than seeking a single name to cover both cases.
 * A specification should first and foremost be clear and precise.  Smooshing together two cases makes specifications less clear, or requires a comment at each use to indicate which of the meanings is intended.
 * We should not assume that every tool will want to handle the cases in the same way.  Some tools may already have more precise handling of some cases, such as the Checker Framework does for `Map.get`.  But I would argue this even if I didn't have a concrete example to hand, just based on getting burned by similar assumptions in the past.

**Comment by wmdietlGC** [comment]
#27 discussed `@LessEnforced` as a separate concept from `@NullnessUnspecified`.

**Comment by cpovirk** [comment]
If we can't find a name that serves both use cases, we should also consider shifting the focus to what you call the "unspecifiable" case. My thinking:

- The case of "I don't want to annotate this yet" might come up a lot when users don't want to annotate whole methods or classes (for which users would use declaration annotations), but...
- I don't think users will commonly need to annotate only _some_ of the types _in a single method signature_ (for which they would use type annotations). (And if they do, we might supply them with ways to do that besides by using a `@NullnessUnspecified` type annotation.)

So maybe we end up with a declaration annotation like `@AnnotatedForNullness(false)` but a type-use annotation like `@NullnessUnspecifiable`.

**Comment by mernst** [comment]
It sounds like you are in agreement that we need to distinguish the cases of:
 * I don't want to annotate this yet.
 * I have considered this method and its semantics cannot be precisely captured by the currnt version of Jspecify.

For the former, a declaration rather than type annotation for this is a nice solution.

For the latter, just "unspecifiable" is a misleading name.  The semantics is not unspecifiable, it's just unspecifiable **by this version of Jspecify**.

**Comment by cpovirk** [comment]
Switching gears:

I forget whether we have previously discussed using `@Nullable` for this but with an element/method/value, as in `@Nullable(weWouldLikeToDoBetterThanOutrightNullableButJspecifyIsLettingUsDown = true)`.

The reason that I thought of this is that it could help with #100 (by making `@Nullable @NullnessUnspecified` impossible). But I don't think that should be a significant motivation.

What we _might_ like more is that it "hides" the "unspecified" case better than a top-level annotation does. After all, we expect that case to be rare. And users who _do_ need that case have surely already seen `@Nullable`, so they might understand this new annotation as "nullable, _but ..._." We can still hold out hope that all we _really_ need to provide is `@DefaultNonNull` and `@Nullable`: Such a simple API makes it easy for users to get started, and it gives us some breathing room to add other nullness annotations down the line (`@PolyNull`? `@EnsuresNonNull`?).

On the other hand, putting the "unspecified" case into `@Nullable` is in some ways the _opposite_ of hiding it: I would have to look into how IDEs handle autocompletion for annotations, but we might be inadvertently thrusting the "unspecified" case in front of users who are just trying to type `@Nullable`.

(Another way to attempt the "hiding" goal (but not to help with #100) is to make this third annotation a nested type enclosed by `@Nullable`: `@Nullable.WeWouldLikeToDoBetterThanOutrightNullableButJspecifyIsLettingUsDown`. This might be better or worse than the approach of using an annotation element, depending on IDE autocompletion, Javadoc rendering, and probably other factors.)

**Comment by cpovirk** [comment]
If we want to think about an annotation element or nested type, we should also think more about whether we would use those features for possible future additions -- [like using it for `@PolyNull`](https://github.com/jspecify/jspecify/issues/79#issuecomment-613024491).

Would we want it to be possible to express `onlyJointly` _and_ `weWouldLikeToDoBetterThanOutrightNullableButJspecifyIsLettingUsDown` _together_? Probably not. That's an argument for not doing _both_ as annotation elements. Not that anyone was proposing doing even _one_ of them that way until just now :)

**Comment by kevin1e100** [comment]
While I understand the desire to allow documenting "why" something is
annotated @NullnessUnspecified, I have to say I don't relish the idea of
"hardcoding" various reasons into the annotation language.  There's the
obvious problem that we may not cover all reasonable reasons, and then
people will still need to use something.  An additional concern of mine is
that having an annotation like this at all is somewhat surprising/confusing
already, and I personally basically hope most people don't have to know
about it; having multiple annotations for the same thing seemingly just
calls more (undue) attention to it.  It seems that comments can be used to
express intent, or there was at some point the idea of giving type-use
annotations an explanation string "value" attribute (i.e.,
@NullnessUnspecified("TODO: look at this more closely").

Regarding making "unspecified" a flavor of @Nullable, I think that was the
idea behind @LessEnforced.  It seemingly has appeal b/c "strict" tools may
be able to "just ignore" the less enforced part and
treat @NullnessUnspecified as just @Nullable.  I think there were other
reasons against this, but one issue with that is that we may get into hot
water with our "unchecked conversions" that we want tools to apply.  For
instance, if a base class method has a parameter
annotated @Nullable(butDontTellMe=true), and I override it with @NonNull,
then we want that to be legal (with optional warning), but a tool that
ignores "butDontTellMe" would flag it as a violation of behavioral
subtyping.  That is to say, it appears we need *all* tools to do special
things for "unspecified" sometimes (though those cases are admittedly rare,
as most of it falls under checker-specific discretion), and to me it seems
the best way to get that is by having a separate annotation for it.

**Comment by cpovirk** [comment]
Switching gears again: When will users first be forced to understand unspecified nullness? Often, it will happen when they encounter an error message that mentions it. For example:

`error: incompatible types: @Nullable Map<@NullnessUnspecified Foo, @NullnessUnspecified Bar> cannot be converted to Map<Foo, Bar>`

In Kotlin-like terms, that's saying:

`Type mismatch: inferred type is Map<Foo!, Bar!>? but Map<Foo, Bar> was expected`

The appearance of `@NullnessUnspecified` in error messages changes how I feel about the long name:

- Before: If people should rarely write `@NullnessUnspecified` in _source code_, [we probably don't mind making them write a long name](https://github.com/jspecify/jspecify/issues/32#issuecomment-541115352).

- After: If people may see `@NullnessUnspecified` frequently in _error messages_, then we may worry more about the length. In particular, I'm sad that _`@NullnessUnspecified` is so noisy when it is never the cause of the failure_ -- at least in "lenient mode." Look back at the error message above: `@NullnessUnspecified` is the most prominent but least important part! (That's arguably made even sadder by the fact that the token "`@NullnessUnspecified`" never appears in the source file I'm imagining.)

Possible solutions (some of which require action only by tool authors, not spec authors):

1. Lenient-mode tools could hide `@NullnessUnspecified` from their output altogether. The downside: This will eventually confuse someone: "Why can I return a `@Nullable T` from this method that returns `T` \[which secretly means `@NullnessUnspecified T`\] but not from this other method that returns `T` \[which means 'just `T`'\]?" This confusion would not exist if we were honest and said "`@NullnessUnspecified T`" in the first place. Similarly, users may be confused when they get a `NullPointerException` from dereferencing an object that their tools say is of type "`String`" (when the tools mean "`@NullnessUnspecified String`"). `@NullnessUnspecified` is already complicated, given that it's the default for unannotated code, so I worry about making it _more_ complicated.

2. Tools could print error messages in "Kotlin style" -- that is, by printing `!` and `?` instead of annotations. The downside: This style is less familiar to some users (though more familiar to others). And somehow, users need to understand the mappings `@Nullable <-> ?` and `@NullnessUnspecified <-> !` (unlike in pure Kotlin, which uses `?` for both source code and error messages). (Additionally, I don't love the use of `!` for "unspecified": To me, `!` suggests "not null" (for which Kotlin uses `!!`). Maybe tools would pick a different symbol (`~`, `¿`, ...) but at the cost of familiarity to Kotlin users.)

3. We could pick a shorter name for the annotation. The downside: Not only is the name less clear in source, it is less clear in error messages -- which, again, is when many users will encounter it for the first time.

Another thing to recall is that we haven't fully settled whether a `@NullnessUnspecified` type annotation will even exist. Suppose for a moment that it does not:

- In one way, that simplifies things: We can say whatever we want in error messages, since there will be no equivalent source code that we would want them to correspond to.

- In another way, it complicates things: No matter what we pick for our error messages (`@NullnessUnspecified`, `unspecified`, `[@Nullable?]`, `!`, ...), users who want to understand it can no longer just look up the class with that name. (And note that symbols like `?` are particularly hard to query for in your favorite search engine. Tools can at least try to help by linking to useful documentation.)

**Comment by kevin1e100** [comment]
> `@NullnessUnspecified` is so noisy when it is never the cause of the failure
Thanks for observing this.

> To me, `!` suggests "not null"
That is in fact what it meant in Spec# I believe.  I've at times used `!?`, FWIW, but am not claiming it's a great solution.

It may not be popular, but besides `String!?` there does seem to be another reasonable solution: just `String` could be used to mean `@NullnessUnspecified String` / `String!?` in error messages as long as both _other_ options, `@NonNull String` / `String!!` and `@Nullable String` / `String?`, are always made explicit.  This would remove "unspecified" clutter as desired and is consistent with the observation that

> the token "`@NullnessUnspecified`" never appears in the source file I'm imagining.

I understand that this would take us away from Kotlin's error messages and is incompatible with the desire to generally write `String!!` as `String`. Instead, this solution would leverage "unspecified" as the default-default (so should still be reasonably intuitive to Java users).
 
The other thing I'll say is that while good error messages will be important, I suspect it's probably out of scope for us to specify error messages exactly.  So while I support the effort to make sure tools can produce reasonable messages, I also think we want to be careful not to wade too far into the area: seems likely to me that tools will want some room for experimentation.

**Comment by cpovirk** [comment]
On the topic of using symbols for types, I opened #129 and left [a reply there](https://github.com/jspecify/jspecify/issues/129#issuecomment-700210356).

**Comment by kevinb9n** [comment]
Reopen if we reverse course on #137.
